### ISSUE TITLE
NO_ERROR while uploading via app

### ISSUE THREAD

**Issue by mirisbowring** [docspell/android-client#51]
Hi,

i generated such QR code as docspell source and scanned the second one (API) with my smartphone - i downloaded the latest version from fdroid.

unfortunately, i am getting the following error:

![Screenshot_20220225_100722_org docspell docspellshare](https://user-images.githubusercontent.com/15715080/155687308-94fed92d-3523-4696-b082-b4225b470562.jpg)

Everything else seems to work. Tried from different smartphones in different connections.

Do i have to take care on something special at my reverse proxy? Websockets are enabled and HTTPS also. All Smartphones are running at least Android 10

**UPDATE**
I checked the server logs - nothing is appearing there.

Unfortunately, i cannot try the setup without reverse Proxy since HTTP ist blocked by Network Policy. Docspell-Share shows this error :`android cleartext communication not permited by network security policy`

According to the AndroidManifest.xml HTTP should be allowed:

https://github.com/docspell/android-client/blob/184f610f7074cb0a675498a15bf35d970438168f/app/src/main/AndroidManifest.xml#L16

**Comment by eikek** [comment]
Oh really strange! I haven't seen this yet, so I have no idea from the top of my head. I'm not aware to do anything special. It doesn't use websockets, only a "plain old" http post request.

If you want to try without reverse proxy, you need to install the previous version. For some reason, enabling custom certificates broke it working in cleartext (although this flag is true).

If nothing can be seen in the logs, it might not even reach the server - is there any indication in the reverse proxy logs that the request hits the server at all? If not, this is quite hard to debug then. You would need to checkout the code and attach a debugger. I only have Android 9 and 11 devices here I think.

**Comment by mirisbowring** [comment]
Thanks for the ideas, i will test and reply on monday.

**Comment by mirisbowring** [comment]
Well, I tried version 0.5 to bypass the secure Communication restriction for access without reverse Proxy and it throws the same error. Version 0.4 fixed this Problem.

With v0.4 I am able to upload a document to 192.168.xxx.xxx:PORT via HTTP (no reverse Proxy). With Revers Proxy i get the stream connection error.

Therefore It's proven, that i am having some troubles with my reverse Proxy, though i am not sure, what. Same config is working for nextcloud e.x.

Are You Using TCP Streams or something like this?

**Comment by mirisbowring** [comment]
Output from AndroidStudio, while debugging:

```
E/upload: Error uploading!
    okhttp3.internal.http2.StreamResetException: stream was reset: NO_ERROR
        at okhttp3.internal.http2.Http2Stream.checkOutNotClosed$okhttp(Http2Stream.kt:646)
        at okhttp3.internal.http2.Http2Stream$FramingSink.flush(Http2Stream.kt:576)
        at okio.ForwardingSink.flush(ForwardingSink.kt:32)
        at okhttp3.internal.connection.Exchange$RequestBodySink.flush(Exchange.kt:228)
        at okio.RealBufferedSink.flush(RealBufferedSink.kt:269)
        at org.docspell.docspellshare.http.HttpRequest$1.writeTo(HttpRequest.java:139)
        at okhttp3.MultipartBody.writeOrCountBytes(MultipartBody.kt:157)
        at okhttp3.MultipartBody.writeTo(MultipartBody.kt:93)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:59)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
        at org.docspell.docspellshare.http.HttpRequest.execute(HttpRequest.java:58)
        at org.docspell.docspellshare.http.UploadManager$UploadWorker.run(UploadManager.java:49)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
```

**Comment by eikek** [comment]
Thanks for the stack trace!! That looks strange. Only thing that came to mind is that I'm reading `http2` in the stack trace. Maybe the app is trying with http2 and the reverse proxy has problems with that? That would explain when curl is using http1.1…? Just a guess…. It could be that okhttp changed some settings in a newer version. 

I'm also using the newest version of the app with an nginx reverse proxy without problems…. I'm not aware that I use anything special in the app. It's just okhttp post request. But I just checked my nginx config and it is configured to support http2.

**Comment by mirisbowring** [comment]
I tried this fix:
https://github.com/square/okhttp/issues/3955#issuecomment-952433744
And the error is gone. Though i get another strange bug:

The Reported HTTP Status Code is `403`:
https://github.com/docspell/android-client/blob/348e7d6b889a5be95d5a8adff1c21cf381f454a1/app/src/main/java/org/docspell/docspellshare/http/UploadManager.java#L49

Though, i don't even see anything in my nginx access logs from this upload request.
Seems like the Request is being intercepted somewhere...

INFO: AndroidStudio is complaining about not closing the response body in https://github.com/docspell/android-client/blob/348e7d6b889a5be95d5a8adff1c21cf381f454a1/app/src/main/java/org/docspell/docspellshare/http/HttpRequest.java#L59

You should return the Response Object instead, read the code from Response Object in https://github.com/docspell/android-client/blob/348e7d6b889a5be95d5a8adff1c21cf381f454a1/app/src/main/java/org/docspell/docspellshare/http/UploadManager.java#L50 and do "response.close()"afterwards.

**Comment by mirisbowring** [comment]
Well, I found the reason for the 403:
Somehow My OPNSense WAF intercepts the Request. Strange behaviour is, that the IP get's banned immitiately after sending the request, tough i did not enable the WAF for this specific IP.

**Comment by eikek** [comment]
Ah nice! So we are one step further :) Thanks for pointing at the bug with not closing the response - not well done!

Was almost hitting ENTER on some thoughts to the 403, which are now obsolete - nice find!

**Comment by mirisbowring** [comment]
And i found the Problem for Block

The Android Client is not sending a correct User-Agent. The default from okhttp3 is not sufficient and will be blocked by OPNSense/PFSense per default.

I've created a UserAgentInterceptor and will create a PR for you to merge.

This Change does also Fix my HTTP/2 Error. Obviously HTTP/2 could not be established since the IP has been banned

**Comment by eikek** [comment]
Thanks! I'm using OPNSense myself, but not their WAF (have never seen it, is this a plugin?). Do you know what user-agent values are blocked by default? Would it then be ok to set something like "Docspell-Android/$version"?

Regarding the NO_STREAM error: was this due to http2 vs http1? I thought the 403 from the waf was coming once okhttp was set to not use http2?

**Comment by mirisbowring** [comment]
The NO_STREAM Error was due to the banning.

Last Year I switched from HAProxy to Nginx due to the NAXSI availability.
Therefore, the IP got banned at the second, the Request entered the WAF.
Therefore, the Server did not established a HTTP/2 session and OKHTTP3 threw this error.

Switching to HTTP/1.1 "fixed" the error because the Handshake is different and the client received the 403 - Banned Message. -> Upload still not working

Those Agents are on the blocklist (blocking all default agents to prevent possible Bots/Script-Kiddies.
```Python-urllib|Nmap|python-requests|libwww-perl|MJ12bot|Jorgee|fasthttp|libwww|Telesphoreo|A6-Indexer|ltx71|okhttp|ZmEu|sqlmap|LMAO/2.0|ltx71|zgrab|Ronin/2.0|Hakai/2.0```

With the #57 User Agent Fix, the HTTP/1.1 Downgrade is not neccessary anymore.


I also added the `Request.close()` call i mentioned above.
